### ISSUE TITLE
Allow relay to work with a self hosted instance, like puppet bolt or puppet agent

### ISSUE THREAD

**Issue by guillemlc** [puppetlabs/relay#87]
## Use Case

1. Lots of people, me included, will not see the value of having to login to a 3rd party service just to use a command line tool.
I can use Terraform from the command line from ANYWHERE. Why would I have to login to your service to use it?

I understand the commercial intention of it, but, in my opinion is a bad idea. I would rather pay you for support and advanced features than just having to pay to connect to 3rd party host in order to do stuff. It is not even the price, that is very fair, it is just that it does not make sense to do it.

2. If you are behind a firewall, or an isolated network, this becomes useless. This is only going to work in always online hosts.

## Describe the Solution You Would Like

If you offer a community edition, I would suggest cut down features ( or better, allow users to get extra modules in a manual way). If your product is valuable and the extra features and support are worth it, people will go for it.

I would like you to go for the same model as Hashicorp Terraform.

## Describe Alternatives You've Considered

We use Terraform on a daily basis. Relay could be an interesting option as we also use Puppet. But with a subscription based command line? No, thank you.

## Additional Context

Aside from the annoyance and technical reasons, I personally find it obnoxious that we have tools going for monthly payment service. If you want to make money, design a good set of features for you tool, training plans, and support. But having users captive under a subscription is not something I am going for.

**Comment by kenazk** [comment]
Hey there. Thanks for the feedback! Just so I understand, are you looking at Relay in context of setting up Terraform in a CI/CD-like workflow? 

Our goal with Relay is to provide a service that can automatically kick off workflows and take action based on 3rd party triggers coming in from your monitoring tool or cloud provider etc. The CLI that you're describing (housed in this repo) is just one part of the overall Relay product meant specifically for workflow administration. The actual _service_ part of Relay that is being offered under a subscription includes the services to listen and respond to triggers from the 3rd party services, creating and running of the actual workflows, maintaining a module ecosystem, and providing 24/7 technical support.  Agreed if it were just a CLI – would be nuts to charge a subscription for that. 

Also, I totally get that you may want to run that kind of service in your own environment, but lots of people have told us they don't want that maintenance overhead. However, we've been exploring open sourcing some of the underlying Relay components that, to your point, provide a stripped down version of Relay that enables folks to run it themselves. That work is going on in this repo: https://github.com/puppetlabs/relay-core. 

On the connectivity piece, we mainly see Relay connecting to APIs (e.g. datadog api) or cloud endpoints (e.g. the AWS management API) and relying upon tools like Puppet or Bolt to provide the last mile configuration management of VMs. But as you mentioned, we also heard from lots of people who have more stringent connectivity requirements. In response, we created the [puppetlabs/relay](https://forge.puppet.com/modules/puppetlabs/relay) module that provides an agent that can run behind a firewall and proxy requests to the VMs without requiring line of sight connectivity.  

In any case , all of that may leave you unconvinced :) would be happy to chat more about what would make Relay valuable for you.

**Comment by guillemlc** [comment]
> Hey there. Thanks for the feedback! Just so I understand, are you looking at Relay in context of setting up Terraform in a CI/CD-like workflow?
> 
> Our goal with Relay is to provide a service that can automatically kick off workflows and take action based on 3rd party triggers coming in from your monitoring tool or cloud provider etc. The CLI that you're describing (housed in this repo) is just one part of the overall Relay product meant specifically for workflow administration. The actual _service_ part of Relay that is being offered under a subscription includes the services to listen and respond to triggers from the 3rd party services, creating and running of the actual workflows, maintaining a module ecosystem, and providing 24/7 technical support. Agreed if it were just a CLI – would be nuts to charge a subscription for that.
> 
> Also, I totally get that you may want to run that kind of service in your own environment, but lots of people have told us they don't want that maintenance overhead. However, we've been exploring open sourcing some of the underlying Relay components that, to your point, provide a stripped down version of Relay that enables folks to run it themselves. That work is going on in this repo: https://github.com/puppetlabs/relay-core.
> 
> On the connectivity piece, we mainly see Relay connecting to APIs (e.g. datadog api) or cloud endpoints (e.g. the AWS management API) and relying upon tools like Puppet or Bolt to provide the last mile configuration management of VMs. But as you mentioned, we also heard from lots of people who have more stringent connectivity requirements. In response, we created the [puppetlabs/relay](https://forge.puppet.com/modules/puppetlabs/relay) module that provides an agent that can run behind a firewall and proxy requests to the VMs without requiring line of sight connectivity.
> 
> In any case , all of that may leave you unconvinced :) would be happy to chat more about what would make Relay valuable for you.

No I do not want Relay to setup Terraform. I think you did not read my Enhancement request or you are replying in an arbitrary way.

I want to NOT have to have an account with Relay. In order to use Relay.

The same way I do not need an account to use Puppet or Terraform or most similar tools.

**Comment by guillemlc** [comment]
I understand what you say, but Terraform does that without an API, you get the binary that contains all that code and you run it.

Lets imagine I want to interact with AWS.

With Terraform I put the AWS credentials ( of course! ) and I run Terraform commands with it. That is it. If I want to also use GPC or Cloudflare or another API, I add the credentials for those APIs.

Maybe one of the APIs is local like VSphere or something I interact with via credentials. maybe a local Kubernetes cluster.

With Relay, I have to be connected to relay.sh, that is adding credentials and then also do all the above, so I have my code flying through a 3rd party server somewhere. 

I do not see the added value that is worth that. Plus, in most enterprise environments that premise would not fly. maybe some places do not care, but most or at least some do.
---

The bottom line is:  I wanted to consider Relay.sh as an alternative to Terraform but I can't as it forces us to log in to a remote server in order to operate, which is unacceptable.

**Comment by kenazk** [comment]
Hi there, sorry for misunderstanding the first time. Definitely not trying to reply arbitrarily. You mentioned a couple times wanting to consider Relay as an alternative to Terraform. What is it about Terraform that is problematic for you today? 

I hear what you're saying about wanting a standalone binary to run Relay. That's not a deployment scenario we currently have planned – but not to say we couldn't look into it in the future.
